### PR TITLE
Using IIS Modules with ASP.NET Core

### LINKED ISSUE #2367: New Topic: Using IIS Modules with ASP.NET Core
This question has reoccurred to me due to Response Compression, Response Caching, and URL Rewriting middlewares hitting the scene.

Some IIS modules can process requests/responses with the reverse-proxy setup and others don't work. For example, IIS Dynamic Compression (`DynamicCompressionModule`) works, while IIS Static Compression (`StaticCompressionModule`) doesn't work.

Should we provide a pair of lists in the Publish to IIS doc (or the ANCM Configuration Reference doc) that shows which IIS modules work and which ones don't work?

Reference List: https://www.iis.net/learn/get-started/introduction-to-iis/iis-modules-overview

### PR DESCRIPTION
**Pull request by GuardRex**
Fixes #2367 

Notes:
* **Client Certification Mapping Authentication** - Does it work? It would take me a bit of time to set that up for testing, so I hope one of the gurus can confirm that it :sparkles: just works :sparkles:.
* **Custom Tracing** - I can't confirm it works, because I don't quite see where that data is going. Is there a trace ETW somewhere? I think I'm not looking in the right place. See: https://github.com/aspnet/AspNetCoreModule/issues/65
* **Disabling IIS modules**- I have info on disabling modules in a couple of middleware docs, but that content should be pulled and those docs should link to this. This is a complete and clear explanation.
* **Minimal module configuration** - AFAICT from testing, the Anonymous Authentication Module is required with the ANCM to make a Core app work. Dev team can let me know if there's more to the story. I bet pulling the unneeded modules has a nice perf story, but unless good perf testing is performed, I'm not in a position to say here one way or the other. I don't have a test rig that can figure that out.

### CODE REVIEW

**Comment by GuardRex** [comment]
@Rick-Anderson @tdykstra This one is ready for technical help/review. Here are the main issues in addition to any that the tech review turns up:

1. **Client Certification Mapping Authentication** - In theory, it should :sparkles: just work :sparkles:. It's difficult for me to test this one here. Is it known to work?
2. **Custom Tracing** - I can't confirm it works, because I don't see where that data is going. Is there a trace ETW somewhere? I think I'm not looking in the right place. See: aspnet/AspNetCoreModule#65
3. **Minimal module configuration** - AFAICT from testing, the Anonymous Authentication Module is the only module along with the ANCM required to make a Core app work. Is there more to the story?

**Comment by Rick-Anderson** [comment]
@shirhatti  can you answer these questions from our `#1` contributor?

**Comment by shirhatti** [comment]
>Client Certification Mapping Authentication - Does it work? It would take me a bit of time to set that up for testing, so I hope one of the gurus can confirm that it ✨ just works ✨.

Client cert should just work, but I haven't used it recently. If you want I can share instructions on how to test to client cert.

> Custom Tracing - I can't confirm it works, because I don't quite see where that data is going. Is there a trace ETW somewhere? I think I'm not looking in the right place. See: aspnet/AspNetCoreModule#65

Yup. That's on my plate. Not familiar of the top of my head and I haven't had a chance to do my homework yet. I promise I'm getting to it soon

> Minimal module configuration - AFAICT from testing, the Anonymous Authentication Module is required with the ANCM to make a Core app work. Dev team can let me know if there's more to the story.

Yes, seeing a 401 is expected behavior.

> I bet pulling the unneeded modules has a nice perf story, but unless good perf testing is performed, I'm not in a position to say here one way or the other. I don't have a test rig that can figure that out.

We plan to do that. It's part of the work captured in aspnet/Performance/issues/186

**Comment by GuardRex** [comment]
Thanks @shirhatti 

### Custom Tracing
I didn't see the traces or data logged anywhere, so I'll wait until I hear back from you on that one. I'll close https://github.com/aspnet/AspNetCoreModule/issues/65 when there's an answer.

### Minimal module configuration
Good ... it should be fine the way that I have it in the doc. We can take another look :eye: when this doc goes into its proper review.

### Client Certification Mapping Authentication
I hope I don't sound tooooo lazy, but I just wanted to avoid the overhead of having to set it up merely for testing if it's already known to work. I was planning to just mark the chart *Yes ... it works* ... if someone (e.g., you) knows that's the case.

Do you feel confident in your prediction, or do you think I should set it up and test it just to make sure?

I have the instructions: https://www.iis.net/configreference/system.webserver/security/authentication/iisclientcertificatemappingauthentication

**Comment by Rick-Anderson** [comment]
>Do you feel confident in your prediction, or do you think I should set it up and test it just to make sure?

I would go with his prediction.

**Comment by GuardRex** [comment]
> Client cert should just work, but I haven't used it recently.

Ok ... I just didn't know if that was strong enough.

The only thing left to address is the custom tracing bit, and we'll see how that turns out.

**Comment by Rick-Anderson** [comment]
@GuardRex  let me know when this is ready to merge.

**Comment by GuardRex** [comment]
@Rick-Anderson To finish it, I'm just waiting to hear back from [@]shiratti on whether or not the Custom Tracing module will work.

Also note that even after I hear back from him that the whole doc will need a final pass (by me) then both a proper technical review and a UE review.

**Comment by Rick-Anderson** [comment]
@GuardRex  maybe we should merge and you could open a issue with that question.

**Comment by GuardRex** [comment]
Yes ... but let me make one final pass on it. I need to check grammar/style. Plus, I should just pull that one module (Custom Tracing) out of the table temporarily, then [@]shirhatti can get back to me later.

Give me perhaps three hours. I'll ping you back here when ready to go.

**Comment by GuardRex** [comment]
Naaaah ... I'll do it now. These docs I'm working on can wait. Stand-by ... I'll have it ready in a few minutes.

**Comment by Rick-Anderson** [comment]
@GuardRex  no hurry really.

**Comment by GuardRex** [comment]
@Rick-Anderson :rocket: ... If you're good on UE, **it's ready**. We can add this bit back later ...
```
**Custom Tracing**<br>`TracingModule` | ? | https://github.com/aspnet/AspNetCoreModule#65
```
The middleware docs discuss module deactivation. I'll get a PR in soon (after this goes live) to remove that content and point those docs here, where the info on module deactivation is much better. :+1:

**Comment by GuardRex** [comment]
:grin: ....... "no hurry" ... you ask me about something and I **am in a hurry**. lol

**Comment by Rick-Anderson** [comment]
Need to add this topic to "aspnetcore\hosting\toc.md" 
@tdykstra  why does the hosting folder have it's own TOC?

**Comment by GuardRex** [comment]
@Rick-Anderson Do you want me to do that?

File: aspnetcore/hosting/iis-modules.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+RoleManager | No | 
+ScriptModule-4.0 | No | 
+Session | No | [Session Middleware](xref:fundamentals/app-state)
+UrlAuthorization | No | 
+UrlMappingsModule | No | [URL Rewriting Middleware](xref:fundamentals/url-rewriting)
+UrlRoutingModule-4.0 | No | [ASP.NET Core  Identity](xref:security/authentication/identity)
+WindowsAuthentication | No | 
+
+## IIS Manager application changes
+When using IIS modules with .NET applications, keep in mind that when you interact with the IIS Manager to configure settings that you're directly changing the *web.config* file of the app. If you later deploy a *web.config* with the app, any changes you made in the IIS Manager on the server will be overwritten by the settings in the deployed *web.config* file. After you've made changes to an app's *web.config* on the server, mirror those changes in your project's *web.config* immediately.
```
Comment (Rick-Anderson) [inline aspnetcore/hosting/iis-modules.md]: ~~When using IIS modules with .NET applications, keep in mind that~~

Comment (Rick-Anderson) [inline aspnetcore/hosting/iis-modules.md]: When you use IIS Manager to configure settings, you're directly ...

Comment (Rick-Anderson) [inline aspnetcore/hosting/iis-modules.md]: If you deploy your app and include *web.config*, any changes you made with IIS Manger will be overwritten by the deployed *web.config* file.

Comment (Rick-Anderson) [inline aspnetcore/hosting/iis-modules.md]: If you make changes to the servers *web.config* file, copy the updated *web.config file to ....

**Comment by Rick-Anderson** [comment]
@GuardRex  Let's get confirmation from @tdykstra  that's the right approach. I think it's pretty simple to add the TOC entry

**Comment by GuardRex** [comment]
@Rick-Anderson Based on his response, I'll add the entry (or not) and fix :point_up: those probs.

If it's ok (and he comments here tonight), I'll do it tomorrow morning. I'm working on a `dotnet/docs` thingy right now.

**Comment by tdykstra** [comment]
@Rick-Anderson @GuardRex The hosting and publishing TOC nodes will be combined into one TOC node, and when that is done later this week or early next week there won't be separate toc.md files.  For now, adding to aspnetcore\hosting\toc.md is fine.

**Comment by GuardRex** [comment]
@Rick-Anderson @tdykstra Ok ... I'll add it as part of the updates that @Rick-Anderson wants. I'll get on it this afternoon.

**Comment by GuardRex** [comment]
@Rick-Anderson @tdykstra I'm past the mega-PR over there now, I'll be working this tonight and get 'er done! :grinning:

**Comment by GuardRex** [comment]
@Rick-Anderson It's in the TOC ... https://github.com/GuardRex/Docs/blob/guardrex/iis-modules/aspnetcore/hosting/toc.md

**Comment by GuardRex** [comment]
Should I pull the author byline on this ... isn't that the way things are headed?

**Comment by GuardRex** [comment]
@Rick-Anderson ok ... ready for :eyes: ... let me know if you want the author byline removed.

**Comment by bangbingsyb** [comment]
Looks like the "Minimum module configuration" in this doc caused a lot of confusing among users - while it is the right minimum requirement to make ASP.NET core app work, people think it's the fastest configuration and use it in all sorts of perf tests:

https://github.com/aspnet/IISIntegration/issues/747
https://github.com/aspnet/Performance/issues/186

I think it might worth to make some clarifications in this doc

**Comment by guardrex** [comment]
Let's discuss further over on https://github.com/aspnet/Docs/issues/5844.

**Pull request merged**